finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,628 @@
+
+    fn addresses() -> Vec<Multiaddr> {
+        (0..NUM_NODES)
+            .map(|id| {
+                ScMultiaddr::empty()
+                    .with(ScProtocol::Ip4(Ipv4Addr::new(192, 168, 1, id)))
+                    .with(ScProtocol::Tcp(30333))
+                    .with(ScProtocol::P2p(ScPeerId::random().into()))
+            })
+            .collect()
+    }
+
+    async fn keyboxes() -> Vec<KeyBox> {
+        let num_keyboxes: usize = NUM_NODES.into();
+        let keystore = Arc::new(KeyStore::new());
+        let mut auth_ids = Vec::with_capacity(num_keyboxes);
+        for _ in 0..num_keyboxes {
+            let pk = keystore.ed25519_generate_new(KEY_TYPE, None).await.unwrap();
+            auth_ids.push(AuthorityId::from(pk));
+        }
+        let mut result = Vec::with_capacity(num_keyboxes);
+        for i in 0..num_keyboxes {
+            result.push(KeyBox::new(
+                NodeIndex(i),
+                AuthorityVerifier::new(auth_ids.clone()),
+                AuthorityPen::new(auth_ids[i].clone(), keystore.clone())
+                    .await
+                    .expect("The keys should sign successfully"),
+            ));
+        }
+        result
+    }
+
+    async fn build() -> (Discovery, Vec<SessionHandler>) {
+        let mut handlers = Vec::new();
+        for (keybox, address) in keyboxes().await.into_iter().zip(addresses()) {
+            handlers.push(
+                SessionHandler::new(keybox, SessionId(43), vec![address.into()])
+                    .await
+                    .unwrap(),
+            );
+        }
+        (Discovery::new(Duration::from_millis(MS_COOLDOWN)), handlers)
+    }
+
+    #[tokio::test]
+    async fn broadcasts_when_clueless() {
+        let (mut discovery, mut handlers) = build().await;
+        let handler = &mut handlers[0];
+        let mut messages = discovery.discover_authorities(handler);
+        assert_eq!(messages.len(), 1);
+        let message = messages.pop().unwrap();
+        assert_eq!(
+            message,
+            (
+                DiscoveryMessage::AuthenticationBroadcast(handler.authentication()),
+                DataCommand::Broadcast
+            )
+        );
+    }
+
+    #[tokio::test]
+    async fn requests_from_single_when_only_some_missing() {
+        let num_nodes: usize = NUM_NODES.into();
+        let (mut discovery, mut handlers) = build().await;
+        for i in 1..num_nodes - 1 {
+            let authentication = handlers[i].authentication();
+            assert!(handlers[0].handle_authentication(authentication));
+        }
+        let handler = &mut handlers[0];
+        let messages = discovery.discover_authorities(handler);
+        assert_eq!(messages.len(), 2);
+        for message in messages {
+            assert!(
+                matches!(message, (DiscoveryMessage::Request(node_ids, authentication), DataCommand::SendTo(_, _)) if node_ids == vec![NodeIndex(6)] && authentication == handler.authentication())
```

Comment:
uhm, please wrap such long lines :| clippy/fmt have no power within macros

Reply:
Ouch, I missed that, bleh, fixing.

Reply:
Fixed, although it seems fmt has some power somewhere, only not consistently? Strange anyway.